clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -467,7 +467,8 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
         auto Kind =
             E->isIncrementOp() ? cir::UnaryOpKind::Inc : cir::UnaryOpKind::Dec;
         // NOTE(CIR): clang calls CreateAdd but folds this to a unary op
-        value = emitUnaryOp(E, Kind, input);
+        value =
+            emitUnaryOp(E, Kind, input, /*nsw=*/type->isSignedIntegerType());
```

Comment:
Classic codegen doesn't add `nsw` here. Based on the checks above, we should only get here if `E->canOverflow()` is false. For instance, I think we hit this code with this:

```
short f(short s) {
  s++;
  return s;
}
```
In this case, `s` is promoted to int before the operation, so it can't overflow, and classic codegen doesn't set the `nsw` flag.

https://godbolt.org/z/P4a11n4v5